### ISSUE TITLE
Hello，I want to know if this code can be directly used in AIS data streams?

### ISSUE THREAD

**Issue by Tagore-Ye** [valschmidt/cpa#1]
Hello，I want to know if this code can be directly used in AIS data streams?
Below is the def of you code:
def init(self,length,x,y,speed,heading):
self.length = length
self.pos = np.array([x,y])
self.speed = speed
self.heading = heading

If the initial parameters of the ship are directly substituted with latitude, longitude, speed and heading in AIS data, is the CPA calculated at last still accurate?

**Comment by valschmidt** [comment]
No. The positions of both your ship and the AIS contact must be converted to a "projected" coordinate system in which x and y are in meters (or yards or whatever). 

But here's a receipt that while not perfect, will suffice. The idea here is to create a temporary cartesian coordinate system centered on your own vessel, and then calculate the coordinate of the AIS contact in meters assuming a spherical earth (which works pretty well for contacts that aren't too far away). Here's some pseudo code that uses the fact that there are 1852 meters per minute of change in latitude, and that the true distance represented by a change in longitude decreases with the cosine of latitude on a spherical Earth (because meridians converge at the poles), to calculate the difference in x and y of two vessels in meters where:

LATos, LONos: Latitude and longitude of your own ship in decimal degrees
LATc, LONc: Latitude and Longitude of the AIS contact in decimal degrees
```
dx = (LATc - LATos) * 60 min/deg * 1852 m/deg
dy = (LONc - LONos) * cos(LATos x pi/180) * 60 min/deg * 1852 m/deg
```

With this then the CPA calculation would look something like:

```
Vos  = Vessel(0,0,speed,heading)
Vc = Vessel(dx,dy,speed,heading)
(tcpa,Vesselcpa,OScpa,Rcpa,Bcpa) = Vos.cpa(Vc)
```

`Vesselcpa` provides a tuple of the coordinates of the other's vessel position at CPA. If you want this in lat/lon, put these values in for dx and dy in the equation above and solve for LATc,LONc. 

Hope this helps.

**Comment by Tagore-Ye** [comment]
> No. The positions of both your ship and the AIS contact must be converted to a "projected" coordinate system in which x and y are in meters (or yards or whatever).
> 
> But here's a receipt that while not perfect, will suffice. The idea here is to create a temporary cartesian coordinate system centered on your own vessel, and then calculate the coordinate of the AIS contact in meters assuming a spherical earth (which works pretty well for contacts that aren't too far away). Here's some pseudo code that uses the fact that there are 1852 meters per minute of change in latitude, and that the true distance represented by a change in longitude decreases with the cosine of latitude on a spherical Earth (because meridians converge at the poles), to calculate the difference in x and y of two vessels in meters where:
> 
> LATos, LONos: Latitude and longitude of your own ship in decimal degrees LATc, LONc: Latitude and Longitude of the AIS contact in decimal degrees
> 
> ```
> dx = (LATc - LATos) * 60 min/deg * 1852 m/deg
> dy = (LONc - LONos) * cos(LATos x pi/180) * 60 min/deg * 1852 m/deg
> ```
> 
> With this then the CPA calculation would look something like:
> 
> ```
> Vos  = Vessel(0,0,speed,heading)
> Vc = Vessel(dx,dy,speed,heading)
> (tcpa,Vesselcpa,OScpa,Rcpa,Bcpa) = Vos.cpa(Vc)
> ```
> 
> `Vesselcpa` provides a tuple of the coordinates of the other's vessel position at CPA. If you want this in lat/lon, put these values in for dx and dy in the equation above and solve for LATc,LONc.
> 
> Hope this helps.

Thanks for your suggestions!Another question ,as we konw,AIS speed is measured in knots.and this speed can be directly substituted into your formula?Or could you recommend me some documents or source code for calculating CPA based on AIS data? Thanks a lot！

**Comment by valschmidt** [comment]
Hm. I should document this better - speed should expressed in m/s. 1 knot is about 0.5 m/s.

**Comment by Tagore-Ye** [comment]
> Hm. I should document this better - speed should expressed in m/s. 1 knot is about 0.5 m/s.

got it!THX
